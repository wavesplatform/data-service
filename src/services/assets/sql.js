const pg = require('knex')({ client: 'pg' });
const { compose } = require('ramda');

const mget = ids =>
  pg('assets')
    .select('*')
    .whereIn('asset_id', ids)
    .toString();

module.exports = {
  get: id => mget([id]),
  mget,
  search: ({ ticker }) => {
    const filter = q => {
      if (ticker === '*') return q.whereNotNull('ticker');
      else return q.where('ticker', ticker);
    };

    return compose(
      q => q.toString(),
      filter
    )(pg('assets').select('*'));
  },
};
