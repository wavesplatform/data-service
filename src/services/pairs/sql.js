const pg = require('knex')({ client: 'pg' });

/** query :: Array[Object] -> String */
const query = pairs =>
  pg({ t: 'pairs' })
    .select('*')
    .whereIn(
      ['t.amount_asset_id', 't.price_asset_id'],
      pairs.map(pair => [pair.amountAsset, pair.priceAsset])
    )
    .toString();

/** search :: Object -> String */
const search = ({ limit }) =>
  pg({ t: 'pairs' })
    .select('*')
    .orderByRaw('t.volume_waves desc NULLS LAST')
    .limit(limit)
    .toString();

module.exports = {
  /** get :: Object -> String */
  get: pair => query([pair]),
  /** mget :: Array[Object] -> String */
  mget: query,
  search,
};
