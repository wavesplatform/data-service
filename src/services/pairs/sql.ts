import * as knex from 'knex';
import { compose } from 'ramda';
import { escapeForTsQuery, prepareForLike } from '../../utils/db';
import {
  PairsGetRequest,
  PairsMgetRequest,
  PairsSearchRequest,
  SearchByAssetRequest,
  SearchByAssetsRequest,
} from '.';
import { AssetIdsPair } from '../../types';

const pg = knex({ client: 'pg' });

const COLUMNS = [
  'first_price',
  'last_price',
  'volume',
  'quote_volume',
  'high',
  'low',
  'weighted_average_price',
  'txs_count',
  'volume_waves',
];

const isSearchByAssetRequest = (
  searchByAssetRequest: PairsSearchRequest
): searchByAssetRequest is SearchByAssetRequest => {
  return (
    typeof searchByAssetRequest === 'object' &&
    searchByAssetRequest !== null &&
    searchByAssetRequest.hasOwnProperty('search_by_asset')
  );
};

const isSearchByAssetsRequest = (
  searchByAssetRequest: PairsSearchRequest
): searchByAssetRequest is SearchByAssetsRequest => {
  return (
    typeof searchByAssetRequest === 'object' &&
    searchByAssetRequest !== null &&
    searchByAssetRequest.hasOwnProperty('search_by_assets')
  );
};

const getMatchExactly = (matchExactly: boolean[] | undefined): boolean[] => {
  const DEFAULT_MATCH_EXACTLY = false;

  if (typeof matchExactly === 'undefined') {
    return [DEFAULT_MATCH_EXACTLY, DEFAULT_MATCH_EXACTLY];
  } else {
    if (matchExactly.length === 1) {
      return [matchExactly[0], matchExactly[0]];
    } else {
      return matchExactly;
    }
  }
};

const query = (pairs: AssetIdsPair[], matcher: string): string =>
  pg({ t: 'pairs' })
    .select(COLUMNS)
    .select({
      amount_asset_id: 'aa.asset_id',
      price_asset_id: 'pa.asset_id',
    })
    .join({ aa: 'assets' }, 'aa.uid', 't.amount_asset_uid')
    .join({ pa: 'assets' }, 'pa.uid', 't.price_asset_uid')
    .leftJoin(
      { matcher_addr: 'addresses' },
      'matcher_addr.uid',
      't.matcher_address_uid'
    )
    .whereIn(
      ['aa.asset_id', 'pa.asset_id'],
      pairs.map(pair => [pair.amountAsset, pair.priceAsset])
    )
    .whereIn('matcher_address_uid', function() {
      this.select('uid')
        .from('addresses')
        .where('address', matcher)
        .limit(1);
    })
    .toString();

const searchAssets = (
  qb: knex.QueryBuilder,
  query: string,
  matchExactly: boolean,
  tableAlias: string = 't'
) => {
  // will be used on searchable_asset_name search
  const cleanedQuery = escapeForTsQuery(query);
  return qb.distinct('asset_uid').from(function(this: knex.QueryBuilder) {
    this.table({ [tableAlias]: 'assets' })
      .column({ asset_uid: `${tableAlias}.uid` })
      .where(`${tableAlias}.asset_id`, query)
      .orWhere(
        `${tableAlias}.ticker`,
        'ilike',
        prepareForLike(query, { matchExactly })
      )
      .unionAll(q =>
        q
          .from({ [`${tableAlias}2`]: 'assets_metadata' })
          .column({ asset_uid: `${tableAlias}2.asset_uid` })
          .where(
            `${tableAlias}2.asset_name`,
            'ilike',
            prepareForLike(query, { matchExactly })
          )
      )
      .unionAll(q =>
        compose((q: knex.QueryBuilder) =>
          cleanedQuery.length
            ? q.orWhereRaw(
                `${tableAlias}3.searchable_asset_name @@ to_tsquery(?)`,
                [`${cleanedQuery}:*`]
              )
            : q
        )(
          q
            .from({ [`${tableAlias}3`]: 'assets' })
            .column({ asset_uid: `${tableAlias}3.uid` })
            .where(
              `${tableAlias}3.asset_name`,
              'ilike',
              prepareForLike(query, { matchExactly })
            )
        )
      )
      .as('a');
  });
};

export const get = (r: PairsGetRequest): string => query([r.pair], r.matcher);
export const mget = (r: PairsMgetRequest): string => query(r.pairs, r.matcher);
export const search = (req: PairsSearchRequest): string => {
  // asset - prefix search of amount or price assets
  // asset1/asset2 - prefix search of amount asset by asset1
  //                 and prefix search of price asset by asset2
  //                 or amount asset by asset2 and price asset by asset1

  const { matcher, limit } = req;
  const q = pg({ t: 'pairs' })
    .select(COLUMNS.map(column => `t.${column}`))
    .select({
      amount_asset_id: 'aa.asset_id',
      price_asset_id: 'pa.asset_id',
    })
    .join({ aa: 'assets' }, 'aa.uid', 't.amount_asset_uid')
    .join({ pa: 'assets' }, 'pa.uid', 't.price_asset_uid')
    .orderByRaw('volume_waves desc NULLS LAST')
    .whereIn('matcher_address_uid', function() {
      this.select('uid')
        .from('addresses')
        .where('address', matcher)
        .limit(1);
    })
    .limit(limit);

  if (isSearchByAssetRequest(req)) {
    const { search_by_asset: amountAsset, match_exactly: matchExactly } = req;
    return q
      .with('assets_cte', (qb: knex.QueryBuilder) =>
        searchAssets(qb, amountAsset, getMatchExactly(matchExactly)[0])
      )
      .join(
        { amountCte: 'assets_cte' },
        'amountCte.asset_uid',
        't.amount_asset_uid'
      )
      .join(
        { priceCte: 'assets_cte' },
        'priceCte.asset_uid',
        't.amount_asset_uid'
      )
      .toString();
  } else if (isSearchByAssetsRequest(req)) {
    const [amountAsset, priceAsset] = req.search_by_assets;

    const [amountAssetExactly, priceAssetExaclty] = getMatchExactly(
      req.match_exactly
    );

    return q
      .with('assets_cte', (qb: knex.QueryBuilder) =>
        qb
          .select({
            amount_asset_uid: 't1.asset_uid',
            price_asset_uid: 't2.asset_uid',
          })
          .from({
            t1: searchAssets(
              pg.queryBuilder(),
              amountAsset,
              amountAssetExactly,
              'a'
            ),
          })
          .crossJoin((qb: knex.QueryBuilder) =>
            searchAssets(qb, priceAsset, priceAssetExaclty, 'p').as('t2')
          )
      )
      .join(
        { direct_cte: 'assets_cte' },
        {
          'direct_cte.amount_asset_uid': 't.amount_asset_uid',
          'direct_cte.price_asset_uid': 't.price_asset_uid',
        }
      )
      .unionAll((qb: knex.QueryBuilder) =>
        qb
          .select(COLUMNS.map(column => `t1.${column}`))
          .select({
            amount_asset_id: 'aa.asset_id',
            price_asset_id: 'pa.asset_id',
          })
          .from({ t1: 'pairs' })
          .join(
            { reverse_cte: 'assets_cte' },
            {
              'reverse_cte.amount_asset_uid': 't1.price_asset_uid',
              'reverse_cte.price_asset_uid': 't1.amount_asset_uid',
            }
          )
          .join({ aa: 'assets' }, 'aa.uid', 't1.amount_asset_uid')
          .join({ pa: 'assets' }, 'pa.uid', 't1.price_asset_uid')
          .whereIn('matcher_address_uid', function() {
            this.select('uid')
              .from('addresses')
              .where('address', matcher)
              .limit(1);
          })
      )
      .toString();
  } else {
    return q.toString();
  }
};
