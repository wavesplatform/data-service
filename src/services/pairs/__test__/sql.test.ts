import { get, mget, search } from '../sql';

describe('sql query from pairs', () => {
  it('should get one pair', () => {
    expect(
      get({
        amountAsset: '111',
        priceAsset: '222',
        matcher: '333',
      })
    ).toMatchSnapshot();
  });

  it('should get many pairs', () => {
    expect(
      mget({
        pairs: [
        {
          amountAsset: '111',
          priceAsset: '222',
        },
        {
          amountAsset: '333',
          priceAsset: '444',
        },
      ], matcher: '555'})
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset', () => {
    expect(
      search({
        search_by_asset: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj',
        matcher: '',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset exactly', () => {
    expect(
      search({
        search_by_asset: '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj',
        match_exactly: [true],
        matcher: '',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for one asset exactly', () => {
    expect(
      search({
        search_by_asset: '¯\\_(ツ)_/¯',
        match_exactly: [true],
        matcher: '',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({ search_by_assets: ['BTC', 'WAVES'], matcher: '', limit: 10 })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({
        search_by_assets: ['¯\\_(ツ)_/¯', 'WAVES'],
        matcher: '',
        limit: 10,
      })
    ).toMatchSnapshot();
  });

  it('should search pairs for two assets (amount and price)', () => {
    expect(
      search({
        search_by_assets: ['¯\\_(ツ)_/¯', 'WAVES'],
        match_exactly: [true, false],
        matcher: '',
        limit: 10,
      })
    ).toMatchSnapshot();
  });
});
