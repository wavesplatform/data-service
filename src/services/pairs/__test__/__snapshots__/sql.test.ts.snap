// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sql query from pairs should get many pairs 1`] = `"select \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" left join \\"addresses\\" as \\"matcher_addr\\" on \\"matcher_addr\\".\\"uid\\" = \\"t\\".\\"matcher_address_uid\\" where (\\"aa\\".\\"asset_id\\", \\"pa\\".\\"asset_id\\") in (('111', '222'), ('333', '444')) and \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '555' limit 1)"`;

exports[`sql query from pairs should get one pair 1`] = `"select \\"first_price\\", \\"last_price\\", \\"volume\\", \\"quote_volume\\", \\"high\\", \\"low\\", \\"weighted_average_price\\", \\"txs_count\\", \\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" left join \\"addresses\\" as \\"matcher_addr\\" on \\"matcher_addr\\".\\"uid\\" = \\"t\\".\\"matcher_address_uid\\" where (\\"aa\\".\\"asset_id\\", \\"pa\\".\\"asset_id\\") in (('111', '222')) and \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '333' limit 1)"`;

exports[`sql query from pairs should search pairs for one asset 1`] = `"with \\"assets_cte\\" as (select distinct \\"asset_uid\\" from (select \\"t\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"ticker\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' union all select \\"t2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' union all select \\"t3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t3\\" where \\"t3\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj%' or t3.searchable_asset_name @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj:*')) as \\"a\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for one asset exactly 1`] = `"with \\"assets_cte\\" as (select distinct \\"asset_uid\\" from (select \\"t\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or \\"t\\".\\"ticker\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' union all select \\"t2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' union all select \\"t3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t3\\" where \\"t3\\".\\"asset_name\\" ilike '7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj' or t3.searchable_asset_name @@ to_tsquery('7FJhS4wyEKqsp77VCMfCZWKLSMuy1TWskYAyZ28amWFj:*')) as \\"a\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for one asset exactly 2`] = `"with \\"assets_cte\\" as (select distinct \\"asset_uid\\" from (select \\"t\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t\\" where \\"t\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"t\\".\\"ticker\\" ilike E'¯\\\\\\\\_(ツ)_/¯' union all select \\"t2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"t2\\" where \\"t2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯' union all select \\"t3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"t3\\" where \\"t3\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯') as \\"a\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"amountCte\\" on \\"amountCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets_cte\\" as \\"priceCte\\" on \\"priceCte\\".\\"asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 1`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_uid\\" as \\"amount_asset_uid\\", \\"t2\\".\\"asset_uid\\" as \\"price_asset_uid\\" from (select distinct \\"asset_uid\\" from (select \\"a\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = 'BTC' or \\"a\\".\\"ticker\\" ilike 'BTC%' union all select \\"a2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike 'BTC%' union all select \\"a3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a3\\" where \\"a3\\".\\"asset_name\\" ilike 'BTC%' or a3.searchable_asset_name @@ to_tsquery('BTC:*')) as \\"a\\") as \\"t1\\" cross join (select distinct \\"asset_uid\\" from (select \\"p\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'WAVES' or \\"p\\".\\"ticker\\" ilike 'WAVES%' union all select \\"p2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'WAVES%' union all select \\"p3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p3\\" where \\"p3\\".\\"asset_name\\" ilike 'WAVES%' or p3.searchable_asset_name @@ to_tsquery('WAVES:*')) as \\"a\\") as \\"t2\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" and \\"direct_cte\\".\\"price_asset_uid\\" = \\"t\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) union all select \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_uid\\" = \\"t1\\".\\"price_asset_uid\\" and \\"reverse_cte\\".\\"price_asset_uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t1\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 2`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_uid\\" as \\"amount_asset_uid\\", \\"t2\\".\\"asset_uid\\" as \\"price_asset_uid\\" from (select distinct \\"asset_uid\\" from (select \\"a\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"a\\".\\"ticker\\" ilike E'¯\\\\\\\\_(ツ)_/¯%' union all select \\"a2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%' union all select \\"a3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a3\\" where \\"a3\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯%') as \\"a\\") as \\"t1\\" cross join (select distinct \\"asset_uid\\" from (select \\"p\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'WAVES' or \\"p\\".\\"ticker\\" ilike 'WAVES%' union all select \\"p2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'WAVES%' union all select \\"p3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p3\\" where \\"p3\\".\\"asset_name\\" ilike 'WAVES%' or p3.searchable_asset_name @@ to_tsquery('WAVES:*')) as \\"a\\") as \\"t2\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" and \\"direct_cte\\".\\"price_asset_uid\\" = \\"t\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) union all select \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_uid\\" = \\"t1\\".\\"price_asset_uid\\" and \\"reverse_cte\\".\\"price_asset_uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t1\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;

exports[`sql query from pairs should search pairs for two assets (amount and price) 3`] = `"with \\"assets_cte\\" as (select \\"t1\\".\\"asset_uid\\" as \\"amount_asset_uid\\", \\"t2\\".\\"asset_uid\\" as \\"price_asset_uid\\" from (select distinct \\"asset_uid\\" from (select \\"a\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a\\" where \\"a\\".\\"asset_id\\" = E'¯\\\\\\\\_(ツ)_/¯' or \\"a\\".\\"ticker\\" ilike E'¯\\\\\\\\_(ツ)_/¯' union all select \\"a2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"a2\\" where \\"a2\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯' union all select \\"a3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"a3\\" where \\"a3\\".\\"asset_name\\" ilike E'¯\\\\\\\\_(ツ)_/¯') as \\"a\\") as \\"t1\\" cross join (select distinct \\"asset_uid\\" from (select \\"p\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p\\" where \\"p\\".\\"asset_id\\" = 'WAVES' or \\"p\\".\\"ticker\\" ilike 'WAVES%' union all select \\"p2\\".\\"asset_uid\\" as \\"asset_uid\\" from \\"assets_metadata\\" as \\"p2\\" where \\"p2\\".\\"asset_name\\" ilike 'WAVES%' union all select \\"p3\\".\\"uid\\" as \\"asset_uid\\" from \\"assets\\" as \\"p3\\" where \\"p3\\".\\"asset_name\\" ilike 'WAVES%' or p3.searchable_asset_name @@ to_tsquery('WAVES:*')) as \\"a\\") as \\"t2\\") select \\"t\\".\\"first_price\\", \\"t\\".\\"last_price\\", \\"t\\".\\"volume\\", \\"t\\".\\"quote_volume\\", \\"t\\".\\"high\\", \\"t\\".\\"low\\", \\"t\\".\\"weighted_average_price\\", \\"t\\".\\"txs_count\\", \\"t\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t\\".\\"price_asset_uid\\" inner join \\"assets_cte\\" as \\"direct_cte\\" on \\"direct_cte\\".\\"amount_asset_uid\\" = \\"t\\".\\"amount_asset_uid\\" and \\"direct_cte\\".\\"price_asset_uid\\" = \\"t\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) union all select \\"t1\\".\\"first_price\\", \\"t1\\".\\"last_price\\", \\"t1\\".\\"volume\\", \\"t1\\".\\"quote_volume\\", \\"t1\\".\\"high\\", \\"t1\\".\\"low\\", \\"t1\\".\\"weighted_average_price\\", \\"t1\\".\\"txs_count\\", \\"t1\\".\\"volume_waves\\", \\"aa\\".\\"asset_id\\" as \\"amount_asset_id\\", \\"pa\\".\\"asset_id\\" as \\"price_asset_id\\" from \\"pairs\\" as \\"t1\\" inner join \\"assets_cte\\" as \\"reverse_cte\\" on \\"reverse_cte\\".\\"amount_asset_uid\\" = \\"t1\\".\\"price_asset_uid\\" and \\"reverse_cte\\".\\"price_asset_uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"aa\\" on \\"aa\\".\\"uid\\" = \\"t1\\".\\"amount_asset_uid\\" inner join \\"assets\\" as \\"pa\\" on \\"pa\\".\\"uid\\" = \\"t1\\".\\"price_asset_uid\\" where \\"matcher_address_uid\\" in (select \\"uid\\" from \\"addresses\\" where \\"address\\" = '' limit 1) order by volume_waves desc NULLS LAST limit 10"`;
